udp/protocol.go

```diff
@@ -0,0 +1,174 @@
+// Copyright 2015 The Chihaya Authors. All rights reserved.
+// Use of this source code is governed by the BSD 2-Clause license,
+// which can be found in the LICENSE file.
+
+package udp
+
+import (
+	"bytes"
+	"encoding/binary"
+	"net"
+
+	"github.com/chihaya/chihaya/stats"
+	"github.com/chihaya/chihaya/tracker/models"
+)
+
+// initialConnectionID is the magic initial connection ID specified by BEP 15.
+var initialConnectionID = []byte{0, 0, 0x04, 0x17, 0x27, 0x10, 0x19, 0x80}
+
+// eventIDs maps IDs to event names.
+var eventIDs = []string{"", "completed", "started", "stopped"}
+
+var (
+	errMalformedPacket = models.ProtocolError("malformed packet")
+	errMalformedIP     = models.ProtocolError("malformed IP address")
+	errMalformedEvent  = models.ProtocolError("malformed event ID")
+	errBadConnectionID = models.ProtocolError("bad connection ID")
+)
+
+// handleTorrentError writes err to w if err is a models.ClientError.
+func handleTorrentError(err error, w *Writer) {
+	if err == nil {
+		return
+	}
+
+	if _, ok := err.(models.ClientError); ok {
+		w.WriteError(err)
+		stats.RecordEvent(stats.ClientError)
+	}
+}
+
+// handlePacket decodes and processes one UDP request, returning the response.
+func (s *Server) handlePacket(packet []byte, addr *net.UDPAddr) (response []byte, actionName string) {
+	if len(packet) < 16 {
+		return // Malformed, no client packets are less than 16 bytes.
```

Comment:
I decided not to, to avoid making it a target for (admittedly low-efficiency) UDP amplification attacks. I'm not sure if that's really valid though.